Port to 1.4

Pop000100 opened the pull request:
See commit descriptions for a changelogs.

I am going to wait a day to merge this.

Pop000100 committed:
Intial Port to 1.4

Also removed most of the magic numbers that have a Terraria.ID Equivalent

Please note now I have to fix the autofisher.

Pop000100 committed:
Fix somethings, I missed intially

- Autofisher had some issues, as did the reinforced vest.
- Fixed a bug in the chest vacuum(placing on two chests causes a out of bound error)
- and I broke some recipes in the previous commit

gardenappl commented on NPCs/BaseballCaveBat.cs:
```diff
@@ -38,16 +38,12 @@ public override float SpawnChance(NPCSpawnInfo spawnInfo)
-			if(Main.rand.Next(3) == 0)
-				Item.NewItem(npc.position, npc.width, npc.height, ModContent.ItemType<BaseballCap>());
+			npcLoot.Add(ItemDropRule.Common(ItemID.ChainKnife, 250));
+			npcLoot.Add(ItemDropRule.Common(ItemID.DepthMeter, 100));
```
Do we need to add `ItemDropRule`s for vanilla items? I honsetly don't remember why I had that code in the first place.

gardenappl replied:
Wait, I'm dumb, I forgot that these are custom entities and not `GlobalNPC`s, haha.

gardenappl commented on NPCs/BaseballCaveBat.cs:
```diff
@@ -38,16 +38,12 @@ public override float SpawnChance(NPCSpawnInfo spawnInfo)
+			npcLoot.Add(ItemDropRule.Common(ItemID.ChainKnife, 250));
+			npcLoot.Add(ItemDropRule.Common(ItemID.DepthMeter, 100));
+			npcLoot.Add(ItemDropRule.NormalvsExpert(ModContent.ItemType<BaseballBat>(), 4, 2));
+			npcLoot.Add(ItemDropRule.Common(ModContent.ItemType<BaseballCap>(), 3));
```
Is that 50% drop rate in Expert mode? What was I thinking? Maybe I intended this to be 2%, 4%, and 3%.

gardenappl replied:
Then we'd have to swap the NormalvsExpert rule so it's 2% in normal mode and 4% in expert mode.

gardenappl replied:
(whoops, haha)

gardenappl replied:
(forgot this was a rare enemy and not a GlobalNPC for standard bats)

gardenappl commented on Projectiles/AutofisherBobber.cs:
```diff
@@ -51,200 +52,201 @@ public override void SetDefaults()
+				const double ten = 10;
+				Vector2 vector2 = new Vector2((float)(Projectile.position.X + (double)Projectile.width * 0.5), (float)(Projectile.position.Y + (double)Projectile.height * 0.5));
+				//float num5 = (float)(Main.player[Projectile.owner].position.X + (double)(Main.player[Projectile.owner].width / 2) - vector2.X);
+				//float num6 = (float)(Main.player[Projectile.owner].position.Y + (double)(Main.player[Projectile.owner].height / 2) - vector2.Y);
```
I would've just removed the commented-out code if I were you.😅 But this is fine, too.

Pop000100 replied:
I'll remove all the commented out code, I had to make changes to the auto fisher again(fishing incorrect items).

gardenappl commented on Projectiles/AutofisherBobber.cs:
```diff
@@ -51,200 +52,201 @@ public override void SetDefaults()
+							int Type = (int)Projectile.ai[0];
 							Item newItem = new Item();
 							newItem.SetDefaults(Type, true);
 							if(Type == 3196)
```
Might be fun to see the `ItemID` here.

gardenappl replied:
On second thought, after looking at the later autofisher code, I think it's okay to just keep the numbers.

gardenappl commented on Tiles/Autofisher.cs:
```diff
@@ -75,7 +74,7 @@ public override bool NewRightClick(int i, int j)
 			{
 				GoldensMisc.Log(e);
-				Main.NewText("Error! Report this error and send the Terraria/Logs/Logs.txt file!");
+				Main.NewText("Error! Report this error and send the Terraria/Modloader/Logs/Client.log file!");
```
Good catch.

Pop000100 replied:
Thank you

gardenappl commented on Tiles/ChestVacuum.cs:
```diff
@@ -45,21 +44,35 @@ public void SetDefaultsPostContent()
-			Item.NewItem(i * 16, j * 16, 48, 32, mod.ItemType(GetType().Name));
+			Tile tile1 = Main.tile[i, j + 1];
+			Tile tile2 = Main.tile[i + 1, j + 1];
+			if ((tile2 != null || tile2.IsActive) && tile1.frameX != (tile2.frameX - 18))
```
Seems like a null reference could happen here?

Pop000100 replied:
Let me check on a server, tiles can't be null on a client(and single player). This isn't true on a server I spaced that out.

Pop000100 replied:
It looks like this hook is never called on the server so it is safe.

gardenappl replied:
I just meant like
`tile2 != null && tile2.IsActive` 
rather than
`tile2 != null || tile2.IsActive`

gardenappl commented on description.txt:
```diff
@@ -13,4 +13,4 @@ Current features include:
 And a lot more! Go to the forum page for the full list.
  
-Thanks to Zerokk for his help with sprites and to Pop000100 for maintaining the mod.
+Thanks to Zerokk for his help with sprites and to Pop000100 for maintaining the Mod.
```
Why the uppercase?

Pop000100 replied:
That is a mistake I used visual studio to replace all "mod." figuring there would be no collateral damage I was barely wrong.

gardenappl reviewed (commented):
Amazing work! I only have a couple of questions.

Pop000100 committed:
More fixes

- Made the gasterblaster do damage, the summoning changes in 1.4 broke it.
- The mirror was missing sound in the 1.3 version too.
- commented out the mechanic rod tweak since 1.4 made it even more common than that.
- and finally redid the autofisher again since it wasn't producing the right stuff.

Pop000100 committed:
Increment version number

forgot to do it last commit

gardenappl commented:
Sorry to bother you, but how hard would it be to allow the Autofisher to fish up enemies? I remember back in 1.3 I mentioned it would be fun if it could summon Duke Fishron.

Pop000100 commented:
the recent update to the fisher should make it easier, I'll look into it. but it probably should be split from this PR.

I thought about doing it for the bloodmoon enemies but I decided to not since you asked for me to not add or remove content. And I saw you commented out Duke Fishron.

Did you just decide against it or did it have problems?

gardenappl commented:
Well, in 1.3 I thought that automatically summoning Duke Fishron would be just a novelty/easter egg, so I said it would be cool but then just didn't bother. (and after first developing the Autofisher I honestly did not want to touch that code again.)

In 1.4 it seems like fishing enemies plays a slightly bigger role, so now it might be worth looking into.

Pop000100 commented:
> and after first developing the Autofisher I honestly did not want to touch that code again.
I am getting there. Still one more thing to do before its done.
> 
> In 1.4 it seems like fishing enemies plays a slightly bigger role, so now it might be worth looking into.

I'll see about adding it since i have to dive back into the Autofisher anyway.

Pop000100 committed:
The final changes for now

- fix chumbucket
- add the ability to spawn enemies

Pop000100 commented:
Added the ability to spawn enemies for the autofisher including the bloodmoon enemies.

gardenappl commented:
Awesome! I don't know how to thank you for the hard work.

Pop000100 committed:
Change to a standard datatype

- I added a custom data type but it already  existed so a moved back to the standard

Pop000100 commented:
When you have time can you review and approve the changes.

gardenappl commented:
I've been reviewing commits as they came along, though mostly skimming a lot of big changes.

The code *looks* fine, and I assume you tested it.

One tiny concern: is it necessary to use .hjson for localization? Do we need to convert the Russian and Chinese files, too? The Russian lang just needs one simple tweak: replace 30% with 35% for Autofisher tooltip.

Pop000100 commented:
I have tested it and I had someone report some bugs and I gave him a pre-release build, he said it fixed his issues. I can a revert the lang files for now, I don't know when or if they will be ending support for .lang.

EDIT: Seems like they removed .lang support completely.

Yes in that case I need to convert both the chinese and russian localizations

There is a bug in the russian file that crashes the converter tool.
`WorldGen.AncientForge=Добавляем древние кузницы...
WorldGen.AncientForge=Добавляем древние адские кузницы...`

Which one is correct?

gardenappl commented:
Whoops, it should be:
```
WorldGen.AncientForge=Добавляем древние кузницы...
WorldGen.AncientHellforge=Добавляем древние адские кузницы...
```

Pop000100 committed:
Localization Fixes

- Fixes all known issues with the localization
- Converted the other languages

Pop000100 commented:
I'd like to merge it.
Any final thoughts?

gardenappl commented:
I think we're good now, thanks! I will update the forum post later today.

gardenappl commented:
How do mod releases work for 1.4? Is it just in Steam Workshop?
I think maybe you should announce a new release on the forums, and then I'll edit the initial post with updated info.

Pop000100 commented:
Yeah you publish ingame through tmodloader and it pushes to workshop.
Want me to do it?

Edit: I accidentally removed you from contributors trying to see If I could switch roles(I am the owner, you are a contributor, I felt this was inaccurate.)
As a aside do you get pings everytime someone posts on the mods or just me?

Edit 2 : Published the update since I noticed you don't have TModloader installed on steam
I edited my post since I can't double post.

Pop000100 commented:
want me to do it?

gardenappl commented:
I got a ping when you changed my role in the project, other than that, I get nothing. I don't know when the mod updates are published exactly.
